File: kotlin/src/test/kotlin/com/edgeveda/sdk/ChatTypesTest.kt
package com.edgeveda.sdk

import org.junit.Assert.*
import org.junit.Test

/**
 * Unit tests for ChatMessage, ChatRole, and SystemPromptPreset.
 */
class ChatTypesTest {

    // ── ChatMessage ────────────────────────────────────────────────────────────

    @Test
    fun `ChatMessage stores role and content correctly`() {
        val msg = ChatMessage(ChatRole.USER, "Hello world")
        assertEquals(ChatRole.USER, msg.role)
        assertEquals("Hello world", msg.content)
    }

    @Test
    fun `ChatMessage timestamp defaults to current time`() {
        val before = System.currentTimeMillis()
        val msg = ChatMessage(ChatRole.ASSISTANT, "Hi")
        val after = System.currentTimeMillis()
        assertTrue(msg.timestamp.time in before..after)
    }

    @Test
    fun `ChatMessage data class equality holds for same values`() {
        val t = java.util.Date(1_000_000L)
        val a = ChatMessage(ChatRole.USER, "test", t)
        val b = ChatMessage(ChatRole.USER, "test", t)
        assertEquals(a, b)
    }

    @Test
    fun `ChatMessage data class inequality on different roles`() {
        val t = java.util.Date(1_000_000L)
        val a = ChatMessage(ChatRole.USER, "test", t)
        val b = ChatMessage(ChatRole.ASSISTANT, "test", t)
        assertNotEquals(a, b)
    }

    @Test
    fun `ChatMessage data class inequality on different content`() {
        val t = java.util.Date(1_000_000L)
        val a = ChatMessage(ChatRole.USER, "hello", t)
        val b = ChatMessage(ChatRole.USER, "world", t)
        assertNotEquals(a, b)
    }

    @Test
    fun `ChatMessage copy produces independent instance`() {
        val original = ChatMessage(ChatRole.USER, "original")
        val copy = original.copy(content = "copy")
        assertEquals("original", original.content)
        assertEquals("copy", copy.content)
        assertEquals(original.role, copy.role)
    }

    @Test
    fun `ChatMessage accepts empty content`() {
        val msg = ChatMessage(ChatRole.SYSTEM, "")
        assertEquals("", msg.content)
    }

    // ── ChatRole ───────────────────────────────────────────────────────────────

    @Test
    fun `ChatRole has exactly 3 values`() {
        assertEquals(3, ChatRole.entries.size)
    }

    @Test
    fun `ChatRole entries contain SYSTEM USER and ASSISTANT`() {
        val names = ChatRole.entries.map { it.name }.toSet()
        assertEquals(setOf("SYSTEM", "USER", "ASSISTANT"), names)
    }

    @Test
    fun `ChatRole SYSTEM value is lowercase system`() {
        assertEquals("system", ChatRole.SYSTEM.value)
    }

    @Test
    fun `ChatRole USER value is lowercase user`() {
        assertEquals("user", ChatRole.USER.value)
    }

    @Test
    fun `ChatRole ASSISTANT value is lowercase assistant`() {
        assertEquals("assistant", ChatRole.ASSISTANT.value)
    }

    @Test
    fun `ChatRole value returns lowercase of name`() {
        for (role in ChatRole.entries) {
            assertEquals(role.name.lowercase(), role.value)
        }
    }

    // ── SystemPromptPreset ────────────────────────────────────────────────────

    @Test
    fun `SystemPromptPreset Assistant text is non-empty`() {
        assertTrue(SystemPromptPreset.Assistant.text.isNotEmpty())
    }

    @Test
    fun `SystemPromptPreset Coder text is non-empty`() {
        assertTrue(SystemPromptPreset.Coder.text.isNotEmpty())
    }

    @Test
    fun `SystemPromptPreset Concise text is non-empty`() {
        assertTrue(SystemPromptPreset.Concise.text.isNotEmpty())
    }

    @Test
    fun `SystemPromptPreset Creative text is non-empty`() {
        assertTrue(SystemPromptPreset.Creative.text.isNotEmpty())
    }

    @Test
    fun `SystemPromptPreset Custom returns provided prompt as text`() {
        val custom = SystemPromptPreset.Custom("You are a pirate.")
        assertEquals("You are a pirate.", custom.text)
    }

    @Test
    fun `SystemPromptPreset Custom with empty string returns empty text`() {
        val custom = SystemPromptPreset.Custom("")
        assertEquals("", custom.text)
    }

    @Test
    fun `SystemPromptPreset Custom data class equality`() {
        val a = SystemPromptPreset.Custom("same")
        val b = SystemPromptPreset.Custom("same")
        assertEquals(a, b)
    }

    @Test
    fun `SystemPromptPreset Custom data class inequality on different prompts`() {
        val a = SystemPromptPreset.Custom("a")
        val b = SystemPromptPreset.Custom("b")
        assertNotEquals(a, b)
    }

    @Test
    fun `all preset texts are distinct`() {
        val texts = listOf(
            SystemPromptPreset.Assistant.text,
            SystemPromptPreset.Coder.text,
            SystemPromptPreset.Concise.text,
            SystemPromptPreset.Creative.text,
        )
        assertEquals("All preset texts should be distinct", texts.size, texts.toSet().size)
    }
}

File: web/src/__tests__/ChatTypes.test.ts
import { ChatRole, SystemPromptPreset, getSystemPrompt } from '../ChatTypes';

describe('ChatTypes', () => {
  describe('ChatRole enum', () => {
    it('SYSTEM === "system"', () => {
      expect(ChatRole.SYSTEM).toBe('system');
    });

    it('USER === "user"', () => {
      expect(ChatRole.USER).toBe('user');
    });

    it('ASSISTANT === "assistant"', () => {
      expect(ChatRole.ASSISTANT).toBe('assistant');
    });

    it('has exactly 6 string values (including TOOL_CALL, TOOL_RESULT, SUMMARY)', () => {
      const values = Object.values(ChatRole).filter((v) => typeof v === 'string');
      expect(values).toHaveLength(6);
    });

    it('TOOL_CALL === "tool_call"', () => {
      expect(ChatRole.TOOL_CALL).toBe('tool_call');
    });

    it('TOOL_RESULT === "tool_result"', () => {
      expect(ChatRole.TOOL_RESULT).toBe('tool_result');
    });

    it('SUMMARY === "summary"', () => {
      expect(ChatRole.SUMMARY).toBe('summary');
    });
  });

  describe('SystemPromptPreset enum', () => {
    it('ASSISTANT preset exists', () => {
      expect(SystemPromptPreset.ASSISTANT).toBeDefined();
    });

    it('CODER preset exists', () => {
      expect(SystemPromptPreset.CODER).toBeDefined();
    });

    it('CONCISE preset exists', () => {
      expect(SystemPromptPreset.CONCISE).toBeDefined();
    });

    it('CREATIVE preset exists', () => {
      expect(SystemPromptPreset.CREATIVE).toBeDefined();
    });

    it('has exactly 4 string values', () => {
      const values = Object.values(SystemPromptPreset).filter(
        (v) => typeof v === 'string'
      );
      expect(values).toHaveLength(4);
    });
  });

  describe('getSystemPrompt', () => {
    it('returns non-empty string for ASSISTANT', () => {
      const prompt = getSystemPrompt(SystemPromptPreset.ASSISTANT);
      expect(typeof prompt).toBe('string');
      expect(prompt.trim()).not.toBe('');
    });

    it('returns text containing "programmer" for CODER', () => {
      const prompt = getSystemPrompt(SystemPromptPreset.CODER);
      expect(prompt.toLowerCase()).toMatch(/code|programmer/);
    });

    it('returns text containing "concise" or "brief" for CONCISE', () => {
      const prompt = getSystemPrompt(SystemPromptPreset.CONCISE);
      expect(prompt.toLowerCase()).toMatch(/concise|brief/);
    });

    it('returns text for CREATIVE', () => {
      const prompt = getSystemPrompt(SystemPromptPreset.CREATIVE);
      expect(typeof prompt).toBe('string');
      expect(prompt.trim()).not.toBe('');
    });

    it('all presets return distinct prompt texts', () => {
      const prompts = Object.values(SystemPromptPreset)
        .filter((v) => typeof v === 'string')
        .map((preset) => getSystemPrompt(preset as SystemPromptPreset));
      const unique = new Set(prompts);
      expect(unique.size).toBe(prompts.length);
    });

    it('ASSISTANT prompt mentions "assistant"', () => {
      const prompt = getSystemPrompt(SystemPromptPreset.ASSISTANT);
      expect(prompt.toLowerCase()).toContain('assistant');
    });
  });
});
